[NBQian] iP

NBQian opened the pull request:
DUKE

> "Your mind is for having ideas, not holding them." - David Allen

Duke helps you organise different types of tasks to keep you on track. Tasks include:

- Todo: no deadline, finish anytime
- Deadline: a task with a deadline
- Event: a task that has a starting time and an ending time

Duke has the following functions:

1. list: list down all your tasks
2. mark: mark a task as done
3. unmark: mark a task as undone
4. find: find all tasks that contains a key phrase
5. create: adding tasks to your plan
6. delete: ~~remove~~ tasks from your plan

you can check the source code [here](https://github.com/NBQian/ip/commits/master)

for _java programmers_, want to know how the code that runs the program, `run` looks like 😃? Here you are:
```java
    public void run() {
        ui.welcome();
        boolean isExit = false;
        while (!isExit) {
            try {
                String inputLine = ui.readCommand();
                Command c = Parser.parse(inputLine);
                c.execute(tasks, ui, storage);
                isExit = c.isExit();
            } catch (DukeException e) {
                System.out.println(e.getMessage());
            }
        }
    }
```
Features:

- [X] Managing tasks

- [ ] Managing deadlines

- [ ] Reminders (coming soon)

And it is **FREE!**

Commit:
Add Gradle support

Commit:
Level 1

Commit:
Level 2

Commit:
incomplete level 3

Commit:
Completed level 3

Commit:
Completed Level 4

Commit:
Completed A-TextUiTesting

Commit:
Level 5 completed

Commit:
level 6 completed with a new input.txt file

Commit:
Completed adding enum. Updated test file

Commit:
added java doc, updated Task and its subclasses' mark and unmark methods

Commit:
main class: only use an ArrayList to store all tasks

tasks input by the user can only be saved while Duke is running and will be gone once Duke stops running

Create a data folder with a task.txt file in it to permanently store all tasks

Additionally:
1. changed line break style
2. added different greeting messages depending on the existence of data folder and/or tasks.txt file

Commit:
local saved file updated only when Duke terminates

this means locally saved tasks are not constantly updated whenever the list changes

modified Main method such that changes are immediately reflected locally

Commit:
commit a test file otherwise sourcetree gives an error that does not allow me to switch to master branch

Commit:
Task subclasses store time as a String

this is not very flexible in terms of manipulating the way in which date and time is printed.

Change Task subclasses such that they store date and time as java.time.LocalDateTime

Commit:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Task subclasses store time as a String

# Conflicts:
#	src/main/java/Duke.java

Commit:
removed a bug where writing the local file in the wrong format

Commit:
Merge commit '556af3f47a96b32898ab4cdbd65b16486a4871e8'

* commit '556af3f47a96b32898ab4cdbd65b16486a4871e8':
  Add Gradle support

Commit:
changed the data file path

Commit:
OOP without Parse class

Commit:
Completed A-More OOP

Commit:
deleted a useless file

Commit:
removed a bug where "Parse" is used instead of "Parser"

Commit:
Finished A-Package increment, separated different classes into different files and packages.

Commit:
Completed A-JUnit: tested Event and Todo classes using JUnit test

Commit:
Completed A-JavaDoc

Commit:
Finished Level-9 with new class FindCommand Created

Commit:
reformatted all java files using google coding standard.

Commit:
Added java doc

Commit:
Revert "Completed A-JavaDoc"

This reverts commit 6b07de7f569f09afbbfd26634925f1a5176d158a.

Commit:
Revert "Revert "Completed A-JavaDoc""

This reverts commit 13cb2653bcc0573d8e1ffb52d2e5d9d87d659299.

Commit:
Finished A-CodingStandard

Commit:
Completed Level-9 by adding a FindCommand class and a find method in TaskList class.

Commit:
all these files are preventing me from merging until I commit them

Commit:
Merge tag 'A-CodingStandard'

* tag 'A-CodingStandard':
  reformatted all java files using google coding standard.
  Finished Level-9 with new class FindCommand Created
  Completed A-JavaDoc

# Conflicts:
#	.gitignore
#	src/main/java/Duke.java
#	src/main/java/duke/command/AddCommand.java
#	src/main/java/duke/command/ByeCommand.java
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/command/ListCommand.java
#	src/main/java/duke/command/MarkCommand.java
#	src/main/java/duke/command/UnmarkCommand.java
#	src/main/java/duke/parser/Parser.java
#	src/main/java/duke/storage/Storage.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/Todo.java
#	src/main/java/duke/tasklist/TaskList.java
#	src/main/java/duke/ui/Ui.java

Commit:
Completed level 9

Commit:
added java doc for all public classes

Commit:
Merge tag 'A-JavaDoc'

* tag 'A-JavaDoc':
  Revert "Revert "Completed A-JavaDoc""
  Revert "Completed A-JavaDoc"
  Added java doc

Commit:
committing this file so I can merge

Commit:
Merge tag 'A-CodingStandard'

* tag 'A-CodingStandard':
  all these files are preventing me from merging until I commit them
  reformatted all java files using google coding standard.
  Finished Level-9 with new class FindCommand Created
  Completed A-JavaDoc

# Conflicts:
#	.gitignore
#	src/main/java/duke/command/FindCommand.java

Commit:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  added java doc for all public classes

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/duke/command/UnmarkCommand.java
#	src/main/java/duke/storage/Storage.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/tasklist/TaskList.java
#	src/main/java/duke/ui/Ui.java

Commit:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Completed level 9

# Conflicts:
#	.gitignore
#	src/main/java/Duke.java
#	src/main/java/duke/command/AddCommand.java
#	src/main/java/duke/command/ByeCommand.java
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/command/ListCommand.java
#	src/main/java/duke/command/MarkCommand.java
#	src/main/java/duke/command/UnmarkCommand.java
#	src/main/java/duke/parser/Parser.java
#	src/main/java/duke/storage/Storage.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/Todo.java
#	src/main/java/duke/tasklist/TaskList.java
#	src/main/java/duke/ui/Ui.java

Commit:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Finished A-CodingStandard

# Conflicts:
#	src/main/java/duke/command/AddCommand.java
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/command/MarkCommand.java
#	src/main/java/duke/command/UnmarkCommand.java
#	src/main/java/duke/commandtype/CommandType.java
#	src/main/java/duke/dukeexception/DukeException.java
#	src/main/java/duke/parser/Parser.java
#	src/main/java/duke/storage/Storage.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Todo.java
#	src/main/java/duke/tasklist/TaskList.java
#	src/main/java/duke/ui/Ui.java

Commit:
resolved some bugs after merging all three branches

Commit:
Used checkstyle to reformat java files

Commit:
Initial attempt to use JavaFX to run the application.

Things to be fixed:
1. greeting message to include status of loading of local tasks
2. bye command quits the application immediately
3. "Duke" logo seems weird

angjunkang commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,45 @@
+
+    /**
+     * Constructor that passes in the number of the task to be deleted
+     * @param n
```
Perhaps describe what _n_ is here? I noticed the same issues in several other places too like in _MarkCommand_ and etc.

angjunkang commented on src/main/java/duke/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,39 @@
+public class UnmarkCommand extends Command {
+    private int taskNo;
+
+    public UnmarkCommand(int n) {
```
Perhaps add Javadocs for this method? I noticed the same issue in several other places too but having Javadocs in all public methods is an optional task too.

angjunkang commented on src/main/java/duke/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    /**
+     * on execution, mark the target Task in the TaskList as undone and print out customised message
```
Should this be unmark instead of mark?

angjunkang commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,48 @@
+     * @param message a String to be displayed
+     */
+    public void printToFormat(String message) {
+        String lineBreak1 = "-->-->-->-->-->-->-->-->-->-->-->\n    ";
```
Perhaps a more intuitive variable name here? i.e startLine and endLine

angjunkang reviewed (commented):
LGTM, almost. Just a few nits to fix. Overall, I found your code easy to understand. Variable names are mostly intuitive. I noted a few inconsistencies in the Javadocs like capitialization and a few typos.

panpannnnn commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,40 @@
+        try {
+            storage.update(tasks);
+        } catch (IOException e) {
+            System.out.println("failed to update tasks locally: " + e.getMessage());
```
Line break for better readability, happens in other files as well

panpannnnn commented on src/main/java/duke/commandtype/CommandType.java:
```diff
@@ -0,0 +1,8 @@
+/**
+ * a list of valid Commands as enum
+ */
+public enum CommandType {
```
Good use of enums

panpannnnn commented on src/main/java/duke/dukeexception/DukeException.java:
```diff
@@ -0,0 +1,25 @@
+package duke.dukeexception;
+
+/**
+ * class that handles exceptions unique to the Duke application
```
Capitalise first letters to be more formal

panpannnnn commented on src/main/java/duke/tasklist/TaskList.java:
```diff
@@ -0,0 +1,124 @@
+        try {
+            tasks.get(taskNumber - 1).mark();
+        } catch (Exception e) {
+            throw new DukeException(
```
More layman exception message would be nice

panpannnnn reviewed (commented):
Well organised, sorry I don't really understand the extra things you did so no comment on those.

LeirdaKJW commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,30 @@
+ */
+public class ByeCommand extends Command {
+    /**
+     * determines that user wants to exit the program
```
Shouldn't the first letter of the description be in caps?
```suggestion
     * Determines that user wants to exit the program
```

LeirdaKJW commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    /**
+     * on execution, prints goodbye message
```
Same goes for this and some other methods in other classes, 
```suggestion
     * On execution, prints goodbye message
```

LeirdaKJW commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,114 @@
+        this.taskFolder = new File(folderPath);
+    }
+
+    /**
+     * a method that loads the local tasks into the TaskList if they exist. prints customised
+     * messages if the file and/or folder does not exist, and create them accordingly
+     * @return an empty ArrayList of Tasks if no local tasks are found, or an ArrayList
+     *         containing all local tasks if they are found.
+     */
```
You might want to add a @throws comment in this jdoc, 
```suggestion
    /**
     * a method that loads the local tasks into the TaskList if they exist. prints customised
     * messages if the file and/or folder does not exist, and create them accordingly
     * @return an empty ArrayList of Tasks if no local tasks are found, or an ArrayList
     *         containing all local tasks if they are found.
     * @throws ...
     */
```

LeirdaKJW commented on src/main/java/duke/tasklist/TaskList.java:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    /**
+     * @return all the tasks row by row
```
might want to give a brief description of this method.

LeirdaKJW reviewed (commented):
Overall, the structure of the code is clean and readable. Just a few nitpicks few javadocs and capitalizations.

Commit:
finished Level-10

Commit:
Created jar file, minor changes to some classes in accordance to checkstyle

Commit:
MainWindow AddCommand, DeleteCommand, FindCommand, UnmarkCommand do no have assertion to check if the output messages are correct

The lack of assertion can leave inaccurate output messages unchecked

Adding assertion can ensure that the output messages are accurate

Let's java assertion to each class with unique error message should the assertion fail. For example, if the output message of FindCommand is inaccurate, a message "wrong find message" will be output.

Additionally, a get() method is added to TaskList class to facilitate addition of assertion to the above classes.

Commit:
Class Storage

there are redundant code that is no longer useful. For example, the Status attribute and some output messages

the name of the method "update" is not meaningful enough.

readers of the code might get confused when reading the code

removing the unused attribute and output messages increases the readability of the code

Let's remove the redundant lines and refactor the method "update" to "updateLocal"

Class Duke

method "getLoadStatus" is no longer used because the Status attribute in Storage is removed

removing this method prevents error from occurring

Let's remove the unused method getLoadStatus()

Commit:
Classes Duke, Command, DeleteCommand, UnmarkCommand, AddCommand, ByeCommand, FindCommand, MarkCommand, ListCommand, CommandType, DukeException, Parser, Storage, Deadline, Event, Task, Todo, TaskList, Ui.

Header comments are not in accordance with the format specified in the coding standard.

Comments in the wrong format makes it harder for the readers to understand various methods

Let's change the problematic comments in accordance with the format specified in the coding standard

Additionally, unused method "readCommand" in Ui class is removed.

Commit:
Classes MainWindow, AddCommand, DeleteCommand, FindCommand, UnmarkCommand

"==" is used to compare Strings in the java assertions

This may result in the assertion returning false even when the Strings look the same

This can complicate the debugging process when an assertions gives an error

Let's  replace "==" with the .equals() method in java assertions for Strings.

NBQian committed:
Merge pull request #2 from NBQian/branch-A-Assertions

Branch a assertions

NBQian committed:
Merge branch 'master' into branch-A-CodeQuality

NBQian committed:
Merge pull request #3 from NBQian/branch-A-CodeQuality

Branch a code quality

Commit:
initial changes to prepare for C-Tagging

Commit:
Somehow the get method in TaskList magically disappeared after the two branches merged, so I'm adding back that method.

NBQian committed:
Merge pull request #4 from NBQian/branch-A-CodeQuality

Somehow the get method in TaskList magically disappeared after the tw…

Commit:
Further updates to handle C-Tagging

Commit:
Completed C-Tagging by adjusting the initial size of the GUI when launching the app

Commit:
Added comment to give credit to Zhu Yufan

Commit:
Changed font for dialogues, updated credit message in MainWindow.fxml

Commit:
updated catch block in Parser such that all invalid command are caught

Commit:
handles all errors with specific message

Commit:
Now prints tags only if a task have at least 1 tag

Commit:
made sure all files follow the checkstyle, updated the user guide.

Commit:
Cleaned up the code, removed index for the return message of find. Changed font for text at the input window. Updated welcome message

Commit:
updated Ui.png

Commit:
Final touch up of appearance

Commit:
update Ui.png

NBQian committed:
Update README.md

NBQian committed:
Update README.md

NBQian committed:
Update README.md

NBQian committed:
Update README.md
